src/main/java/com/flightstats/hub/dao/aws/ClusterContentService.java

```diff
@@ -143,33 +149,66 @@ public ClusterContentService(
     }
 
     @Override
-    public ContentKey insert(String channelName, Content content) throws Exception {
+    public ContentKey insert(String channelName, Content content) {
+        final Content spokeContent = calculateIndexIfLarge(channelName, content);
+
+        SortedSet<ContentKey> keys = setStableCache(channelName,
+                () -> new TreeSet<>(
+                        Arrays.asList(spokeWriteContentDao.insert(channelName, spokeContent))));
+
+        if (isWriteable(channelName)) {
+            s3SingleWrite(channelName, keys.first());
+        }
+        return keys.first();
+    }
+
+    private Content calculateIndexIfLarge(String channelName, Content content) {
         Content spokeContent = content;
         if (content.isLarge()) {
             s3LargePayloadContentDao.insert(channelName, content);
             spokeContent = largeContentUtils.createIndex(content);
         }
-        final ContentKey key = spokeWriteContentDao.insert(channelName, spokeContent);
-        if (isWriteable(channelName)) {
-            s3SingleWrite(channelName, key);
+        return spokeContent;
+    }
+
+    private SortedSet<ContentKey> setStableCache(String channelName, Supplier<SortedSet<ContentKey>> supplier) {
+        // after stable() seconds, we want to write the cache entry if the write succeeded
+        AtomicReference<ContentKey> ref = new AtomicReference<>();
+
+        ScheduledFuture future = zkCacheStateUpdateExecutor.schedule(() -> {
+            if (ref.get() != null) {
+                latestContentCache.setIfAfter(channelName, ref.get());
+            }
+        }, contentProperties.getStableSeconds() +1, TimeUnit.SECONDS);
+
+        SortedSet<ContentKey> keys;
```

Comment:
Group style choice, but I try to keep variables scoped as tightly as possible. Defining keys outside of the try scope isn't necessary.
```
try {
    SortedSet<ContentKey> keys = supplier.get();
    ref.set(keys.last());
    return keys;
} catch Exception (e) {
    future.cancel(true);
    throw e;
}
```

Reply:
Yep.